adding leica_sep

js731ca opened the pull request:
adding functions and a command to do SEP communication with the MCU to query board-id/-revision/... 

most of the SEP response parsing is recycled from u-boot-imx.git:common/leica_sep.c
providing a script-usable leica_sep cmd instead of hardcoding commands in the board specific files

----
NOTE: this is still a draft/WIP with some TODOs and quirks that need to be figured out and fixed

but the "rough form" stands and works -> looking for constructive criticism :-)

htho-lgs commented on drivers/serial/serial_mxc.c:
```diff
@@ -318,10 +329,36 @@ static int mxc_serial_pending(struct udevice *dev, bool input)
-	.putc = mxc_serial_putc,
-	.pending = mxc_serial_pending,
-	.getc = mxc_serial_getc,
+	.putc = barebox_mxc_serial_putc,
```
Here you change the functionality of `putc` from non-blocking to blocking. I think that serial drivers in u-boot are designed non-blocking. This can have an influence on the "upper" layer.

js731ca replied:
reverted back to non-blocking - with some minor patches to u-boots driver/serial/serial_mxc, which are beeing upstreamed

htho-lgs commented on drivers/serial/serial_mxc.c:
```diff
@@ -318,10 +329,36 @@ static int mxc_serial_pending(struct udevice *dev, bool input)
-	.pending = mxc_serial_pending,
-	.getc = mxc_serial_getc,
+	.putc = barebox_mxc_serial_putc,
+	.pending = barebox_mxc_serial_pending,
```
I believe the difference between RDR in UARTx_USR2 and RXEMPTY in UARTx_UTS is:
- "RxFIFO Empty. Indicates the RxFIFO is empty."
  Okay, RX fifo is empty - without any precondition...
- "Receive Data Ready-Indicates that at least 1 character is received and written to the RxFIFO. If the
URXD register is read and there is only 1 character in the RxFIFO, RDR is automatically cleared."
  We need to receive at least one character, which must not be read, so that this flag is set.

From the function name `pending` I would say the RDR makes more sense.

For TX both could be used:
- "TxFIFO Empty. Indicates that the TxFIFO is empty."
- "Transmitter Complete. Indicates that the transmit buffer (TxFIFO) and Shift Register is empty; therefore
the transmission is complete. TXDC is cleared automatically when data is written to the TxFIFO."

js731ca replied:
thnx; - after comparing both implementations again - with your input in mind - i think i found the issue with the u-boots implementation: the putc path needs to make use of the TXFIFO, otherwise  the message gets "corrupted" = does not make it to the MCU, and therefore no response is created...
-> see follow-up patch

htho-lgs commented on arch/arm/dts/crocodile.dtsi:
```diff
@@ -768,8 +774,10 @@
 		fsl,pins = <
-			CROCODILE_IMX8_UART4_RXD	0x140
-			CROCODILE_IMX8_UART4_TXD	0x140
+			// 0x0 = no pullup/-down, handled on MCU side
```
This is more or less clear, and we don't have any comments on any other value, so I would drop this.

htho-lgs commented on arch/arm/dts/crocodile.dtsi:
```diff
@@ -768,8 +774,10 @@
-			CROCODILE_IMX8_UART4_TXD	0x140
+			// 0x0 = no pullup/-down, handled on MCU side
+			CROCODILE_IMX8_UART4_RXD	0x000
+			// 0x170 = pullup, opendrain, fast-slew-rate, 1x drivestrength
```
Same

js731ca replied:
and removed

htho-lgs reviewed (commented):
Nice solution with the device-tree node and properties!

zandrey commented:
> Nice solution with the device-tree node and properties!

I'd say - exactly the opposite... :disappointed:

DTB shall describe the HW and **not** protocol, which is clearly the case here. Not to mention that schema is missing as a part of this PR, but at this stage I guess it is rather irrelevant.

Moreover: DTBs are synced from Kernel, so this shall either go into Kernel tree first, or put into `-u-boot.dtsi`.

js731ca commented:
and for these reasons it's still a draft-PR -> RFC (-:

the kernel DT was "the first best fit" for a location that is
- not hardcoded in the soucecode, and instead
- linked to the hardware/project/platform description
e.g. does not sit in a common location with per hardware differentiating ifdefs/...
since the available SEP commands (or steps) to identify the hardware and select the appropriate devicetree are or can be (AFAIK) different for each S-:

so should i drop the whole DT/property related part, and we stick to "boot scriptable leica_sep" only?

js731ca committed:
serial: mxc: enable the RX pipeline

on imx8(mm) the RXDMUXSEL needs to be set for data going over the wire
(as observable on a connected 'scope) to actually make it into the
RXFIFO

the reference manual is not overly clear about this, and only
mentiones that "UCR3_RXDMUXSEL should always be set." - and since the
CR3 register reverts to its reset values after setting the baudrate,
setting this bit is done during '_mxc_serial_setbgr'

Signed-off-by: Johannes Schneider <johannes.schneider@leica-geosystems.com>

Series-to: u-boot
Series-cc: bsp-team, festevam@denx.de, peng.fan@oss.nxp.com, sbabic@denx.de, trini@konsulko.com
Series-version: 2

Cover-letter:
serial_mxc: fixing reception

while writing to a imx-serial device is probably thoroughly tested - and obviosly works for the debug-serial - using a serial driver to read data received over the serial interface does not work reliably.

the patches in this series address issues found during the implementation of a custom uboot-command to query a coprocessor, connected to an imx8mm over uart4, for mainboard-identification strings
END

js731ca committed:
serial: mxc: have putc use the TXFIFO

only waiting for TXEMPTY leads to corrupted messages going over the
wire - which is fixed by making use of the FIFO

Signed-off-by: Johannes Schneider <johannes.schneider@leica-geosystems.com>

js731ca committed:
leica_sep: add SEP functions and cmd utility
on crocodile uart4 connects the imx8mm to the coprocessor, which knows
about the board revision/variant. this information is query-able
through 'SEP' commands.

by adding both library functions and a wrapping cmd-line utility these
informations can be used during the boot phase to pick an appropriate
devicetree, or do other configurations the particular MCU firmware
version exposes on its SEP command interface.

Jira: GDH-1129 ("Crocodile: MCU communications in u-boot")
Signed-off-by: Johannes Schneider <johannes.schneider@leica-geosystems.com>

js731ca commented:
@zandrey: the recent rework/forcepush drops the device-node properties with SEP command strings - except for one marker-property in crocodile-u-boot.dtsi; which is needed to "find the one" correct device node... since u-boot does not seem to have a "get_by_label" :-/
so the "which commands to use" would be up to either in the function calls in the boardfile (see example) and/or the bootscript calls to cmd/leica_sep (TBD)
objections? -- or suggestions for a cleaner solution? (-:

htho-lgs commented on arch/arm/dts/crocodile.dtsi:
```diff
@@ -382,7 +382,7 @@
 };
 
-/* LTE module */
+/* MUC/stm32 coprocessor */
```
Changes to this file should come from https://github.com/hexagon-geo-surv/linux-leica/blob/leica/linux-5.19.y/arch/arm64/boot/dts/freescale/crocodile.dtsi

js731ca replied:
ACK - spawned: https://github.com/hexagon-geo-surv/linux-leica/pull/86

htho-lgs commented on arch/arm/dts/crocodile-u-boot.dtsi:
```diff
@@ -68,6 +68,11 @@
 
+&uart4 {
+	// marker for comon/leica_sep.c to find the device-node by
+	leica-sep-node;
```
SEP is a protocol - it should be a marked as PP4 or EFI node

js731ca replied:
ACK, i'll change it to - "leica-MCU-uart"? (even though MCU is... generic; but at least better than the misnomed/confusing "extensible firmware interface"; or the more cryptic PP4 - both leica specific/internal abbreviations)

js731ca committed:
Patch V3
changes:
- reworked comments and config switch-docs to be more verbose
- removed static globals for init_done
- added define for timeout
- moved baudrate into crocodile-uboot.dtsi
- use separate receive buffer in the transfer method, as working buffer to hold the SEP reponse which might be longer than the actually interesting payload - and passed in *data plus data_size

Signed-off-by: Johannes Schneider <johannes.schneider@leica-geosystems.com>

js731ca committed:
Patch V4

Changelog:
- add devicetree binding documantation (in the correct place?)
- fix ifdefs and includes
- get SEP communication during SPL working
- rework kconfig switches, splitting SPL/proper related parts

Signed-off-by: Johannes Schneider <johannes.schneider@leica-geosystems.com>
